File: rds/src/main/java/net/yan100/compose/rds/base/TreeEntity.java
package net.yan100.compose.rds.base;

import com.fasterxml.jackson.annotation.JsonIgnore;
import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.persistence.Column;
import jakarta.persistence.MappedSuperclass;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import net.yan100.compose.core.annotations.BigIntegerAsString;
import net.yan100.compose.core.consts.DataBaseBasicFieldNames;

import java.io.Serial;
import java.io.Serializable;

/**
 * 预排序树
 *
 * @author TrueNine
 * @since 2022-12-12
 */
@Setter
@Getter
@MappedSuperclass
@NoArgsConstructor
@AllArgsConstructor
@Schema(title = "预排序树")
public class TreeEntity extends BaseEntity implements Serializable {
  /**
   * 父id
   */
  public static final String RPI = DataBaseBasicFieldNames.PARENT_ID;
  /**
   * 左节点
   */
  public static final String RLN = DataBaseBasicFieldNames.LEFT_NODE;
  /**
   * 右节点
   */
  public static final String RRN = DataBaseBasicFieldNames.RIGHT_NODE;
  @Serial
  private static final long serialVersionUID = 1L;

  /**
   * 父id
   */
  @JsonIgnore
  @Column(name = DataBaseBasicFieldNames.PARENT_ID)
  @Schema(title = "父id")
  private String rpi;

  /**
   * 左节点
   */
  @JsonIgnore
  @BigIntegerAsString
  @Column(name = DataBaseBasicFieldNames.LEFT_NODE)
  @Schema(title = "左节点", hidden = true)
  private Long rln;

  /**
   * 右节点
   */
  @JsonIgnore
  @BigIntegerAsString
  @Column(name = DataBaseBasicFieldNames.RIGHT_NODE)
  @Schema(title = "右节点", hidden = true)
  private Long rrn;

  /**
   * 节点级别
   */
  @JsonIgnore
  @BigIntegerAsString
  @Schema(title = "节点级别", defaultValue = "0")
  @Column(name = DataBaseBasicFieldNames.NODE_LEVEL)
  private Long nlv = 0L;
}

File: rds/rds-core/src/main/kotlin/net/yan100/compose/rds/core/entity/TreeEntity.kt
package net.yan100.compose.rds.core.entity

import com.fasterxml.jackson.annotation.JsonIgnore
import io.swagger.v3.oas.annotations.media.Schema
import jakarta.persistence.Column
import jakarta.persistence.MappedSuperclass
import net.yan100.compose.core.annotations.BigIntegerAsString
import net.yan100.compose.core.consts.DataBaseBasicFieldNames
import net.yan100.compose.rds.core.annotations.BizCode

/**
 * 预排序树
 *
 * @author TrueNine
 * @since 2022-12-12
 */
@MappedSuperclass
open class TreeEntity : BaseEntity() {
  companion object {
    const val RPI = DataBaseBasicFieldNames.PARENT_ID
    const val RLN = DataBaseBasicFieldNames.LEFT_NODE
    const val RRN = DataBaseBasicFieldNames.RIGHT_NODE
    const val NLV = DataBaseBasicFieldNames.NODE_LEVEL
    const val TGI = DataBaseBasicFieldNames.TREE_GROUP_ID
  }

  /**
   * 父id
   */
  @JsonIgnore
  @Column(name = RPI)
  @Schema(title = "父id")
  open var rpi: String? = null

  /**
   * 左节点
   */
  @JsonIgnore
  @BigIntegerAsString
  @Column(name = RLN)
  @Schema(title = "左节点", hidden = true)
  open var rln: Long? = null

  /**
   * 右节点
   */
  @JsonIgnore
  @BigIntegerAsString
  @Column(name = RRN)
  @Schema(title = "右节点", hidden = true)
  open var rrn: Long? = null

  /**
   * 节点级别
   */
  @JsonIgnore
  @BigIntegerAsString
  @Schema(title = "节点级别", defaultValue = "0")
  @Column(name = NLV)
  open var nlv: Long? = 0L

  /**
   * ### 树组 id，在节点插入时必须更上，在插入时随着父id进行更改
   */
  @BizCode
  @JsonIgnore
  @Schema(title = "树 组id", defaultValue = "0")
  @Column(name = TGI)
  open var tgi: String? = null

  override fun asNew() {
    super.asNew()
    this.rln = 1L
    this.rrn = 2L
    this.nlv = 0
    this.tgi = DataBaseBasicFieldNames.Rbac.ROOT_ID_STR
    this.rpi = null
  }

  override fun toString(): String {
    return withToString(
      super.toString(),
      "rpi" to rpi,
      "rln" to rln,
      "rrn" to rrn,
      "nlv" to nlv,
      "tgi" to tgi
    )
  }
}
